### PR TITLE
Master handle default sequence sht

### PR DESCRIPTION
**Pull request by seb-odoo**
Link to task: https://www.odoo.com/web#id=1853432&action=333&active_id=1251&model=project.task&view_type=form&menu_id=4720

### CODE REVIEW

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 1]
```diff
@@ -1258,8 +1258,62 @@ var BasicModel = AbstractModel.extend({
...
+
+                // We ignore limit.offset because computedList.data will only have the current page elements.
+                var index = Math.min(computedList.data.length - 1, position !== 'top' ? list.limit - 1 : 0);
+                var handleFieldValue = computedList.data[index].data[handleField];
+
+                // This positioning will almost be correct. There might just be an issue if several other lines have the same handleFieldValue.
+                // TODO ideally: if there is an element with the same handleFieldValue,
+                // that one and all the following elements must be incremented by 1 (at least until there is a gap in the numbering).
+                // We don't do it now because it's not an important case.
+                params.force_set_handle_field = true;
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: rename force_set_handle_field to forceSetHandleField

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 2]
```diff
@@ -1258,8 +1258,62 @@ var BasicModel = AbstractModel.extend({
...
+        // We compute the list (get) to apply the pending changes before doing our work.
+        // We use raw: true because we only need to load the first level of relation.
+        var computedList = this.get(list.id, {raw: true});
+
+        var fieldsInfoList = computedList.fieldsInfo.list;
+
+        var hasHandle = false;
+        var handleField = null;
+
+        // We decided to check if there is an handle widget from here instead of pulling it up, to make sure we get the one from the correct view,
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: 80 char width limit. can't even read the end of this line without scrolling

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 3]
```diff
@@ -1258,8 +1258,62 @@ var BasicModel = AbstractModel.extend({
...
+        // and not from the subview or anything else like that.
+        for (var field in fieldsInfoList) {
+            if(fieldsInfoList[field].widget === 'handle') {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: add space between if and (

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 4]
```diff
@@ -1258,8 +1258,62 @@ var BasicModel = AbstractModel.extend({
             parentID: list.id,
             viewType: list.viewType,
         };
+
+        var position = options && options.position || 'top';
+
+        // We compute the list (get) to apply the pending changes before doing our work.
+        // We use raw: true because we only need to load the first level of relation.
+        var computedList = this.get(list.id, {raw: true});
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: no need to compute the list before being certain that there is a handle widget

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 5]
```diff
@@ -1258,8 +1258,62 @@ var BasicModel = AbstractModel.extend({
...
+
+        var fieldsInfoList = computedList.fieldsInfo.list;
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: you can get the same info from list.fieldsInfo I think

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 6]
```diff
@@ -1258,8 +1258,62 @@ var BasicModel = AbstractModel.extend({
...
+
+        var hasHandle = false;
+        var handleField = null;
+
+        // We decided to check if there is an handle widget from here instead of pulling it up, to make sure we get the one from the correct view,
+        // and not from the subview or anything else like that.
+        for (var field in fieldsInfoList) {
+            if(fieldsInfoList[field].widget === 'handle') {
+                if(!hasHandle) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: this if statement seems useless

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 7]
```diff
@@ -1258,8 +1258,62 @@ var BasicModel = AbstractModel.extend({
...
+                    hasHandle = true;
+                    handleField = fieldsInfoList[field].name;
+                    // if there are 2 handle fields on the same list, we take the first one we find (and that will be alphabetically on the field name...)
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: 80 char width limit

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 8]
```diff
@@ -1258,8 +1258,62 @@ var BasicModel = AbstractModel.extend({
...
+                    // TODO: document this behaviour.
+                    break;
+                }
+            }
+        }
+
+        // So if we have an handle widget and at least a line on the list, the currently added element must be at the bottom (or top) of the current page.
+        if (hasHandle && computedList.data.length) {
+            // Also we don't want to force the handle field if the list is not ordered by that field.
+            var isOrderedByHandle = computedList.orderedBy.length && computedList.orderedBy[0].asc === true && computedList.orderedBy[0].name === handleField;
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: 80 char width limit, everywhere

Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: note that you can compute this without using the computedList, but with the list element (so you can compute the computedList later in the next if

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 9]
```diff
@@ -1258,8 +1258,62 @@ var BasicModel = AbstractModel.extend({
...
+
+            if(isOrderedByHandle) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: space between if and (

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 10]
```diff
@@ -1258,8 +1258,62 @@ var BasicModel = AbstractModel.extend({
...
+                // If there are less elements in the list than the limit of the page then take the index of the last existing line.
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: 80 chars limit....

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 11]
```diff
@@ -1258,8 +1258,62 @@ var BasicModel = AbstractModel.extend({
...
+                var index = Math.min(computedList.data.length - 1, position !== 'top' ? list.limit - 1 : 0);
+                var handleFieldValue = computedList.data[index].data[handleField];
+
+                // This positioning will almost be correct. There might just be an issue if several other lines have the same handleFieldValue.
+                // TODO ideally: if there is an element with the same handleFieldValue,
+                // that one and all the following elements must be incremented by 1 (at least until there is a gap in the numbering).
+                // We don't do it now because it's not an important case.
+                params.force_set_handle_field = true;
+                params.handleField = handleField;
+                params.handleFieldValue = (position !== 'top') ? handleFieldValue : handleFieldValue - 1;
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: no need for the parenthesis

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 12]
```diff
@@ -3460,6 +3514,11 @@ var BasicModel = AbstractModel.extend({
                     viewType: params.viewType,
                 });
 
+                // This is a rare case where the defaul_get from the server will be ignored by the view for a certain field.
+                if(params.force_set_handle_field) {
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: space between if and (

File: addons/web/static/src/js/views/basic/basic_model.js [hunk 13]
```diff
@@ -3460,6 +3514,11 @@ var BasicModel = AbstractModel.extend({
                     viewType: params.viewType,
                 });
 
+                // This is a rare case where the defaul_get from the server will be ignored by the view for a certain field.
```
Comment (ged-odoo) [inline addons/web/static/src/js/views/basic/basic_model.js]: 80 chars limit

File: addons/web/static/src/js/views/basic/basic_model.js (hunk 1 above)
Comment (ged-odoo) [reply addons/web/static/src/js/views/basic/basic_model.js]: or maybe overrideHandleField

**Comment by ged-odoo** [comment]
also, now that i think of it, this should be done for the views also (not only the x2manys). If you look at the addDefaultRecord method in the basicmodel, it does basically the same.  Maybe you should extract a function to determine if a list element has a handle field and is sorted by that very field, so you could then use it to make this work for editable lists (stand alone)

**Comment by ged-odoo** [comment]
(not sure)

**Comment by sht-odoo** [comment]
@ged-odoo 

If the standalone lists need to be fixed too, I would prefer to create a separate task if that is ok for you.

And I followed your suggestions regarding the current code. Please let me know if it is ready now.

**Comment by sht-odoo** [comment]
@ged-odoo 

I would prefer to keep this fix simple and to the point, and I propose we create a separate task for the standalone lists if they need to be fixed too.

As for your review, I followed the suggestions, please let me know if it's better/ready now.

**Comment by seb-odoo** [comment]
This has been resolved in https://github.com/odoo/odoo/pull/25483

**Pull request closed without merging**